security-analysis/security-analysis-api/src/main/java/com/powsybl/security/SecurityAnalysisResultBuilder.java

```diff
@@ -94,20 +97,58 @@ public SecurityAnalysisResult build() {
 
         protected final List<LimitViolation> violations = new ArrayList<>();
 
+        protected final Map<LimitViolation, List<Extension<ViolationContext>>> extensionsByViolation = new HashMap<>();
+
         public B setComputationOk(boolean computationOk) {
             this.computationOk = computationOk;
             return (B) this;
         }
 
         public B addViolation(LimitViolation violation) {
             violations.add(Objects.requireNonNull(violation));
+            extensionsByViolation.put(violation, new ArrayList<>());
             return (B) this;
         }
 
         public B addViolations(List<LimitViolation> violations) {
-            violations.forEach(this::addViolation);
+            Objects.requireNonNull(violations).forEach(this::addViolation);
             return (B) this;
         }
+
+        /**
+         * @param limitViolation if it exists already, binding extension to it. Otherwise, it would be added first.
+         * @param extension
+         * @return
+         */
+        public B addViolationContextExtension(LimitViolation limitViolation, Extension<ViolationContext> extension) {
+            Objects.requireNonNull(limitViolation);
+            Objects.requireNonNull(extension);
+            if (!violations.contains(limitViolation)) {
+                this.addViolation(limitViolation);
+            }
+            extensionsByViolation.computeIfAbsent(limitViolation, lv -> new ArrayList<>());
+            extensionsByViolation.computeIfPresent(limitViolation, (lv, list) -> {
+                list.add(extension);
+                return list;
+            });
+            return (B) this;
+        }
```

Comment:
We should not create a ViolationContext each time a violation is created. I think this is the responisibility of the extension to offer a way to retrieve relevant information from the violation (and the current contingency). 

See below for more details.